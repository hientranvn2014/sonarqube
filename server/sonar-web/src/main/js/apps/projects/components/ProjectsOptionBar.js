/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
//@flow
import React from 'react';
import classNames from 'classnames';
import CloseIcon from '../../../components/icons-components/CloseIcon';
import Tooltip from '../../../components/controls/Tooltip';
import PerspectiveSelect from './PerspectiveSelect';
import ProjectsSortingSelect from './ProjectsSortingSelect';
import { translate } from '../../../helpers/l10n';

type Props = {
  onPerspectiveChange: ({ view: string, visualization?: string }) => void,
  onSortChange: (sort: string, desc: boolean) => void,
  onToggleOptionBar: boolean => void,
  open: boolean,
  projects: Array<*>,
  projectsAppState: { loading: boolean, total?: number },
  selectedSort: string,
  view: string,
  visualization?: string
};

export default class ProjectsOptionBar extends React.PureComponent {
  props: Props;

  closeBar = (evt: Event & { currentTarget: HTMLElement }) => {
    evt.currentTarget.blur();
    evt.preventDefault();
    this.props.onToggleOptionBar(false);
  };

  renderSortingSelect() {
    const { projectsAppState, projects, view } = this.props;
    const limitReached =
      projects != null &&
      projectsAppState.total != null &&
      projects.length < projectsAppState.total;

    if (view === 'visualizations' && !limitReached) {
      return (
        <Tooltip overlay={translate('projects.sort.disabled')}>
          <div>
            <ProjectsSortingSelect
              className="projects-topbar-item js-projects-sorting-select disabled"
              onChange={this.props.onSortChange}
              selectedSort={this.props.selectedSort}
              view={this.props.view}
            />
          </div>
        </Tooltip>
      );
    }
    return (
      <ProjectsSortingSelect
        className="projects-topbar-item js-projects-sorting-select"
        onChange={this.props.onSortChange}
        selectedSort={this.props.selectedSort}
        view={this.props.view}
      />
    );
  }

  render() {
    const { open } = this.props;
    return (
      <div className="projects-topbar">
        <div className={classNames('projects-topbar-actions', { open })}>
          <a className="projects-topbar-button button-icon" href="#" onClick={this.closeBar}>
            <CloseIcon />
          </a>
          <div className="projects-topbar-actions-inner">
            <PerspectiveSelect
              className="projects-topbar-item js-projects-perspective-select"
              onChange={this.props.onPerspectiveChange}
              view={this.props.view}
              visualization={this.props.visualization}
            />
            {this.renderSortingSelect()}
          </div>
        </div>
      </div>
    );
  }
}
