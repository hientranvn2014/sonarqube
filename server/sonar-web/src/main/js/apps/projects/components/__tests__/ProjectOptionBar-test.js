/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import { shallow } from 'enzyme';
import ProjectsOptionBar from '../ProjectsOptionBar';
import { click } from '../../../../helpers/testUtils';

it('should render option bar closed', () => {
  expect(shallow(<ProjectsOptionBar open={false} view="overall" />)).toMatchSnapshot();
});

it('should render option bar open', () => {
  expect(
    shallow(
      <ProjectsOptionBar
        open={true}
        view="leak"
        visualization="risk"
        projects={[1, 2, 3]}
        projectsAppState={{ total: 3 }}
      />
    )
  ).toMatchSnapshot();
});

it('should render disabled sorting options for visualizations', () => {
  expect(
    shallow(<ProjectsOptionBar open={true} view="visualizations" visualization="coverage" />)
  ).toMatchSnapshot();
});

it('should call close method correctly', () => {
  const toggle = jest.fn();
  const wrapper = shallow(<ProjectsOptionBar open={true} view="leak" onToggleOptionBar={toggle} />);
  click(wrapper.find('a.projects-topbar-button'));
  expect(toggle.mock.calls).toMatchSnapshot();
});
