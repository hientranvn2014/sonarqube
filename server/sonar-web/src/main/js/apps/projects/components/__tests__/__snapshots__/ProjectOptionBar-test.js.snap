// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should call close method correctly 1`] = `
Array [
  Array [
    false,
  ],
]
`;

exports[`should render option bar closed 1`] = `
<div
  className="projects-topbar"
>
  <div
    className="projects-topbar-actions"
  >
    <a
      className="projects-topbar-button projects-topbar-button-close"
      href="#"
      onClick={[Function]}
    >
      <CloseIcon />
    </a>
    <div
      className="projects-topbar-actions-inner"
    >
      <PerspectiveSelect
        view="overall"
      />
    </div>
  </div>
</div>
`;

exports[`should render option bar open 1`] = `
<div
  className="projects-topbar"
>
  <div
    className="projects-topbar-actions open"
  >
    <a
      className="projects-topbar-button projects-topbar-button-close"
      href="#"
      onClick={[Function]}
    >
      <CloseIcon />
    </a>
    <div
      className="projects-topbar-actions-inner"
    >
      <PerspectiveSelect
        view="visualizations"
        visualization="coverage"
      />
    </div>
  </div>
</div>
`;
